### ISSUE TITLE
Add option to log the database after some time when system is inactive, not KeepassXC

### ISSUE THREAD

**Issue by vpelcak** [keepassxreboot/keepassxc#310]
At the moment there is an option to lock the database after some time of inactivity of the application.

Generally, I use other applications most of the time and need password from time to time (once in few hours). That is enough for the database to lock itself. This kind of defeats the feature of the automatic fill up of the login and passwords, since that requires the database to be open.

If the locking was determined by the (in)activity of the whole system, that would be better, yet, would not leave the database to be vulnerable when not in use.

## Steps to Reproduce (for bugs)

1. Unlock the database
2. Set timeout for locking the database
3. Use other programs, but not the database.
4. Database is locked after timeout

## Your Environment

* KeePassXC version/commit used: (can be found under Help -> About)
* Qt version (e.g. Qt 5.3): Qt 5.7.1
* Compiler (e.g. Clang++3.6.0): gcc 6.3.1+r245113
* Operating System and version: openSUSE Tumbleweed

**Comment by phoerious** [comment]
Duplicate of 134

**Comment by TheZ3ro** [comment]
There is a PR pending #135

**Comment by vpelcak** [comment]
Sorry, but this is not duplicate at all.

I wanted to have the option to choose timeout to be started by the inactivity of the whole system in instead of the inactivity of the KeePassXC itself only.

In other words:
That specified timeout starts when I stop being active on the desktop.
If I set the timeout for 1 hour, lock the screen and then return and unlock the screen after 10 minutes, I still would like to have the database to be open.

**Comment by phoerious** [comment]
Okay, that sounds like a slightly different use case. Such a timeout must not interfere with the "normal" inactivity timeout, because it would basically mean that the DB never locks as long as you are using your PC.

I don't know how difficult this would be to implement on the different platforms.

**Comment by TheZ3ro** [comment]
This seems pretty useless to me because you can instead set a timeout autolock in your OS screensaver and lock KeePassXC with #135

**Comment by vpelcak** [comment]
Well, I'm locking the screen whenever I leave the computer. Once the screen is locked, I don't care if the database is as long as it is closed some reasonable time after I stopped using it to allow it being opened in other location. Yet I would like to have the database not locked while I use the machine and have the computer under control.This thing is completely irrelevant to screen being locked or not. What I care about is,  that the database doesn't lock the file while I use the computer and may potentially need it.Dne 15. 2. 2017 9:14 odpoledne napsal uživatel TheZ3ro <notifications@github.com>:This seems pretty useless to me because you can instead set a timeout autolock in your OS screensaver and lock KeePassXC with #135

**Comment by TheZ3ro** [comment]
So you simply have to disable the KeePassXC timeout and use the screensaver locking

**Comment by vpelcak** [comment]
I also use the database on more machines. Therefore I also do need automatic locking.

**Comment by TheZ3ro** [comment]
I don't know how hard it is to implement a "global" timeout for system inactivity that also is cross-platform

But still I don't understand why you can't set a screensaver for system inactivity. 
At Work you lock your PC and KeePassXC will lock. At Home the screensaver will do the same job as a "global" timeout

**Comment by vpelcak** [comment]
I leave my desk and lock the computer rather often. Therefore locking the database when locking the screen is out of the question for me. I will not enable this option at all (but I understand that some people may need it and will use it). It simply does what I don't need. I want the database to not to lock ideally for the whole time I'm using the computer.

The same for screensaver. I don't need it and don't use it. It brings no benefit for me.

At work I either sit at the computer or leave it for a short while.
When I leave it for longer time, then I'm going home. Therefore 30 minutes system inactivity timeout would correctly guess I left home, and when I am at home, I can open the same database synced over Dropbox and modify it because it locked in meanwhile on my machine at work.

**Comment by phoerious** [comment]
You can always use a database, even when it's locked, as long as you don't write two it from two different locations at the same time.
But I would argue that leaving the DB unlocked and then go home is not the most secure way to handle your password database.

**Comment by vpelcak** [comment]
Well, I know. But I intend to write to it if needed.

Why is it insecure? If the screen is locked and nobody knows my password?

**Comment by phoerious** [comment]
Because it's unencrypted in memory. Someone with physical access to the machine can circumvent the lock screen and get their hands on your passwords.

**Comment by TheZ3ro** [comment]
So at work you leave your PC unlocked with no screensaver/autolock but KeePassXC should lock by inactivity of the whole system?

(At home you lock manually when you leave the desk, ok)

**Comment by vpelcak** [comment]
phoerious - unlikely to happen
TheZ3ro - no

**Comment by mihaiile** [comment]
While I see the user use case I think it is not one that many would use. I am not saying it should not be implemented, but care should be taken not to turn the settings part about the auto locking too confusing, it adds unnecessary complexity to the application, too many settings/options, etc.
I would add this as something like a advanced configuration one would manually add in a configuration file or something.

**Comment by vpelcak** [comment]
Have you ever used Enpass? It has this ability without looking confusing. Sorry, but only you make it overly complicated.

**Comment by phoerious** [comment]
It's not trivial to implement cross-platform. Only because others have done it, it doesn't mean it's easy to implement. And I agree that different settings for different timeouts are confusing. If this gets implemented, we need to be very careful about the settings and how we present them. We don't want users to configure timeouts which don't work as they think they do.

**Comment by vpelcak** [comment]
Naturally. You are the developer, your call.

Thank you for your work. Regardless you implement my request or not.

**Comment by anatoli26** [comment]
This setting was implemented from the beginning in the original KP and the global timeout setting was already requested in #100, with a corresponding printscreen of its implementation in the original KeePass. IMO, those who migrate from KP, won't have any issue with the new setting, moreover, they expect it.

My rationale for this setting is that one may have a short screen lock timeout (say 5 min), so the screen locks quite frequently during the day (say, while at the desk on a phone call, going to WC, etc.), but one would like to lock the KP only after some extended timeout, say 15 min (going to lunch), i.e. leaving the PC for enough time to perform any physical manipulation with it.

Now, this is indeed something not that simple to implement on all platforms. Maybe we could start with Linux, as Windows already has its solution?

**Comment by anatoli26** [comment]
If you are interested in a Linux solution, there's an utility `xprintidle` which prints the time in ms since the last user activity (keyboard/mouse), quite robust, works well on unlocked and locked session. If you inspect its source code (http://archive.ubuntu.com/ubuntu/pool/universe/x/xprintidle/xprintidle_0.2.orig.tar.gz it's just 136 lines .c file) there are basically 3 functions: `XOpenDisplay` > `XScreenSaverQueryExtension` > `XScreenSaverQueryInfo`. Then there's also a workaround for a bug in XScreenSaverExtension when dpms is running, quite trivial too.

So, IMO this option could be introduced conditionally for now (i.e. available on Linux, grayed-out on other platforms), then maybe someone could provide a solution for Windows and OS X.

**Comment by fonic** [comment]
This answer sums up possible solutions for Linux, Mac OS X and Windows:
https://stackoverflow.com/a/21905027

I'd be willing to contribute the Linux part.

**Comment by sergeevabc** [comment]
Frankly, I'm completely puzzled why this function has not yet been implemented since its a common scenario: user switches on the computer, then some apps are launched, one of them is a password manager, it sits in the tray waiting to be unlocked by auto-type shortcut (for example, when user wants to fill in the email credentials), then it stays in unlocked until computers becomes idle for X mins (i.e. 15).

**Comment by UkeFan** [comment]
My Mac Version closes after few minutes of inactiviy no matter what I put at timeouts ):
<img width="743" alt="grafik" src="https://user-images.githubusercontent.com/12803863/60709923-d77c0380-9f11-11e9-97fe-977ad808f69d.png">

**Comment by Barleyman** [comment]
This seems pretty comatose. Re: discussion, it's in original Keepass and not confusing at all. I can see many users wanting to keep the database open while they're using the computer, I certainly do. It's irritating having to type in the DB password over and over. My screen lock timeout is not very long because I don't always remember to lock the PC when I go to the loo etc. 

I guess increasing the screen lock timeout and tying database lock to that would make this behavior less annoying, at least the database won't lock when I'm busy working. But KP behavior of having an explicit "only lock if AFK for xx time" is the best option.

**Comment by phoerious** [comment]
Not sure what you mean. We have options for both locking after some time of inactivity and on lid close/suspend.

**Comment by Barleyman** [comment]
Inactivity of keepasxc, not of the system, which is the whole point of this bug. It'll happily lock every half an hour on the background while I'm working.

**Comment by sergeevabc** [comment]
> Inactivity of keepasxc, not of the system, which is the whole point of this bug…

Dear sir, this issue is named, I quote, “lock the database after some time when system is inactive, not KeepassXC”, whereas you ask for “lock the database after some time when KeepassXC is inactive, not system”. So your request requires another issue to be created, while one seems to be resolved to date.

**Comment by Barleyman** [comment]
> 
> 
> > Inactivity of keepasxc, not of the system, which is the whole point of this bug…
> 
> Dear sir, this issue is named, I quote, “lock the database after some time when system is inactive, not KeepassXC”, whereas you ask for “lock the database after some time when KeepassXC is inactive, not system”. So your request requires another issue to be created and this one seems to be resolved to date.

No I didn't. That behavior is how it works now.

**Comment by droidmonkey** [comment]
The wording of the issue title is rather confusing

**Comment by sergeevabc** [comment]
Hmm, the issue is far from being resolved. For example, I set “Lock databases after inactivity of 30 secs”, then minimize KeepassXC and scroll a webpage with the latest news, so there is an activity, obviously. Yet in 30 secs I see KeepassXC has locked its databases. WTF?!

**Comment by ovz93br43v7** [comment]
My KeepassXC database locks itself automatically while surfing although I have disabled the option autolock at an inactivity of x seconds.

**Comment by GeorchW** [comment]
@phoerious Would you accept a PR on this? I can't test it on all platforms, since I don't own a Mac. But I could probably provide an implementation for Windows and X11.

From the "confusing for users" side of things, my proposal would be to replace the setting

    [x] Lock databases after inactivity of [X sec]

with two settings

    [x] Lock databases when not using KeePassXC for [X sec]
    [x] Lock databases when not using system for    [X sec]

I would be quite surprised if users were puzzled by this, but if that's still confusing, explanatory tooltips could be added.

**Comment by phoerious** [comment]
I don't like adding another option that perhaps 0.01% of our users will actually have a need for. There has to be another solution for this.

**Comment by tsumare** [comment]
@phoerious, I understand that's hyperbole, but do you really think so few people will think "I want to have my passwords available while I'm working, but secure them when I leave my computer"?

This is a distinct concept that doesn't overlap with the existing idle setting. Either you make it an option, or you force it on everyone who prefers the other method. Must we really be so stingy?

"I want my passwords available when I'm here, and secure when I'm not" can hardly be rare.  Does no one work in one application for a while, then later log into another?  Especially with the browser extensions, this is a common workflow.  Make them mutually exclusive if you want, but this option is certainly needed.

**Comment by phoerious** [comment]
I think hardly anybody will care about the distinction between activity in KeePassXC and activity in the system.  The timeout as it is now means it will lock if you haven't actively used KeePassXC within the last X seconds (we may actually add browser extension requests to this @varjolintu). If you want it to stay open indefinitely while you are working, disable the timeout. It will still lock once you lock your screen. We already have four different settings that control how and when KeePassXC shuts the door. I do not fancy adding yet another.

**Comment by sergeevabc** [comment]
+1 for “lock after SYSTEM inactivity”, been waiting for this to happen for God knows how many months.

**Comment by droidmonkey** [comment]
@sergeevabc, do you regularly leave your computer unlocked when not in use? You have not described your specific NEED for this feature based on some sort of requirement in your daily use case. The problem with having two timeouts is that the system timeout effectively nullifies the keepassxc-only timeout. This makes for redundant settings. It also confuses the purpose of lock on screen lock, which is a best practice for security.

**Comment by GeorchW** [comment]
Hmm, if adding a single option for this use case is already too much then I can hardly think of anything that meets your standards. Considering that this issue has 15 upvotes @7.4k github stars, 

Anyway, the thing is that I have a long database password, but a short computer password. When I'm just grabbing a cup of tea at work, I don't want to have to insert my long database password again. So I just lock the computer, then come back a few minutes later, unlock the computer and continue working. If I get caught up in a longer conversation on the way, no problem, my database is automatically locked. 

The second problem is that I'm sometimes called, don't do any input for a few minutes since I'm concentrated on other things, and then my screensaver pops up. Now, this is no problem, the screensaver does not require password entry if interrupted in the first few seconds (i.e. moving the mouse), and even if, the password is short. But if I use the "lock on screensaver" option, then the database instantly locks, and I have to type the long password again.

As long as I don't have this option, I won't set any "user inactivity" or "close on lid" setting, because it is too tedious to type the password all the time, even if I was just gone shortly or was on the phone. So at least in my use case, this setting would improve security.

(On the other hand, I don't really see a use case for "lock after KeePassXC inactivity", so if it was up to me, that setting could be replaced, but it's probably there for a reason.)

**Comment by GeorchW** [comment]
Hmm, if adding a single option for this use case is already too much then I can hardly think of anything that meets your standards. Considering that this issue is one of the more upvoted ones, I don't think that the use case is too rare.

Anyway, the thing is that I have a long database password, but a short computer password. When I'm just grabbing a cup of tea at work, I don't want to have to insert my long database password again. So I just lock the computer, then come back a few minutes later, unlock the computer and continue working. If I get caught up in a longer conversation on the way, no problem, my database is automatically locked.

The second problem is that I'm sometimes called, don't do any input for a few minutes since I'm concentrated on other things, and then my screensaver pops up. Now, this is no problem, the screensaver does not require password entry if interrupted in the first few seconds (i.e. moving the mouse), and even if, the password is short. But if I use the "lock on screensaver" option, then the database instantly locks, and I have to type the long password again.

As long as I don't have this option, I won't set any "user inactivity" or "close on lid" setting, because it is too tedious to type the password all the time, even if I was just gone shortly or was on the phone. So at least in my use case, this setting would improve security.

(On the other hand, I don't really see a use case for "lock after KeePassXC inactivity", so if it was up to me, that setting could be replaced, but it's probably there for a reason.)

**Comment by robsonsobral** [comment]
This is something I miss from KeePass. I bet a lot of people doesn't use it just because they never thought about it. To lock the database after I leave my desk would be awesome.

And I also agree to @GeorchW . "lock after KeePass inactivity" isn't that useful, except if you share a computer. But, in that case, to leave your database unlocked is another kind of problem.

**Comment by phoerious** [comment]
I disagree. KeePassXC inactivity is quite useful if set to a sensible timeout (15-30 minutes). If you don't need a password within this time, the database will lock itself.

**Comment by droidmonkey** [comment]
Thank you for the use case demonstration @GeorchW, that does make sense to me.

**Comment by anatoli26** [comment]
I too was waiting for this option for years. My use-case is opposite to the described one: I have a long delay till the screen locks, but I'd like the KP DB to lock itself after some timeout *shorter* than the system session timeout, but unrelated to the KP usage timeout (the existing option).

**Comment by Barleyman** [comment]
I would imagine for most users current inactivity setting would be the
"wrong" or unexpected behavior. What's the use case for locking the
database if you're actively working on the computer?

Granted, screen saver can mitigate that but as pointed out, maybe you don't
want the database to lock while going to hey coffee..

**Comment by outdooracorn** [comment]
My use case is similar to @GeorchW and I agree with @Barleyman 

My vote is for replacing KeePassXC inactivity with system activity. I think this will be more useful to the majority of users.

**Comment by sergeevabc** [comment]
We ask, we explain, we persuade… damn, for 3 years already… that locking is about locking the app
if user is not present near computer like it is done in the original Keepass, and yet recent 2.6.0b1 is
released with the same misleading, detached-from-reality “lock if app is not used” timer on board. 

![](https://i4.imageban.ru/out/2020/06/26/07e96b6d662d68c18c5c7443e06ec7c6.webp)

**Comment by droidmonkey** [comment]
It's crazy to me that in 3 years time you couldn't teach yourself C++ and write your own implementation.

**Comment by GeorchW** [comment]
@droidmonkey I think that @sergeevabc refers to the fact that most of the time, this feature was not accepted at all. For most of the three years, the maintainers of this project were arguing that they would not accept a PR in this direction (e.g. [here](https://github.com/keepassxreboot/keepassxc/issues/310#issuecomment-611941821)). While I absolutely honor your work and see how you're offended, I think that he does make a valid point.

It's generally unattractive to maintain an own fork of any software: on each update, you have to manually pull and rebuild. This *can* be automated, but is still very tedious and prone to breaking. For security-critical software such as a password manager it might be dangerous to not get any updates for a prolonged period of time. 

(Also, knowing that you just gave something back to the community is a motivation in itself, while the process of getting your PR denied might be frustrating.)

I think a clear statement that a PR is welcome for this feature would help a lot. I don't have the time right now, but I'm sure that, since this is one of the more popular issues (no. 20/500 or so based on upvotes), someone eventually will.

**Comment by droidmonkey** [comment]
No PR for this has been submitted. An issue that is open and marked as new feature is by definition waiting to be implemented.

**Comment by GeorchW** [comment]
> An issue that is open and marked as new feature is by definition waiting to be implemented by anyone.

Huh, okay, didn't realize that. I thought that the issue was open because discussion was encouraged, not because implementation was encouraged. Thanks for clearing that up.

For the record, I did not intend to defend @sergeevabc's tone, which I do consider unacceptable. :-)

**Comment by droidmonkey** [comment]
Yes we have a discussion label for issues that are not pending any implementation

**Comment by SynapseScribe** [comment]
are there any updates? we need this feature because it is way too annoying typing master password every time..

**Comment by WhyNotHugo** [comment]
I'd also very much appreciate this feature. The `Lock databases when session is locked or lid is closed` does not seem to work (at least not on Linux, see https://github.com/keepassxreboot/keepassxc/issues/8194).

I'm currently in a situation where I have to pick between typing the database password all the time, or leaving it permanently unlocked.

**Comment by YannikBartel** [comment]
yeah, i'm still wondering about this behaviour in 2023.

**Comment by WhyNotHugo** [comment]
Is it possible to use the browser-socket to lock the database?

It's trivial for me to write a tiny helper that awaits `n` minutes of idle and tells KPXC to lock itself. I really don't see a need for this to be implemented in-process.

**Comment by michaelk83** [comment]
> Is it possible to use the browser-socket to lock the database?

On Linux you can use DBus: https://github.com/keepassxreboot/keepassxc/wiki/Using-DBus-with-KeePassXC

**Comment by WhyNotHugo** [comment]
> On Linux you can use DBus: https://github.com/keepassxreboot/keepassxc/wiki/Using-DBus-with-KeePassXC

This works, but I'd also need a signal that indicates when a database has been unlocked.

Currently, I can lock the DB after 10 minutes of idle, but then I don't know when to start the next timer, since I don't know when the DB has been unlocked.

**Comment by droidmonkey** [comment]
Does it matter? Just send a lock request when the timer expires due to inactivity. If the database is already locked its a no op. Your timer reset is based on system activity, not database lock status.

**Comment by WhyNotHugo** [comment]
Once I lock the database I have no moment at which I can start the next timer. Starting it immediately would just wake up the system to re-lock a locked database every ten minutes. It's not ideal, but to be honest it should work fine.

**Comment by WhyNotHugo** [comment]
The following works to auto-lock the DB after 10 minutes of idleness. It should work on any wlroots-based compositor.

```sh
#!/bin/sh

while true; do
  wayidle -t 600
  gdbus call --session --dest org.keepassxc.KeePassXC.MainWindow --object-path /keepassxc --method org.keepassxc.KeePassXC.MainWindow.lockAllDatabases
  # TODO: wait for signal indicating that the db is unlocked again.
done
```

**Comment by nykoo** [comment]
This lacking future of lock timeout based on overall system activity could be a main reason for me to return to keepass original... :(
Really hop this will be implemented.

**Comment by Migsi** [comment]
> Once I lock the database I have no moment at which I can start the next timer. Starting it immediately would just wake up the system to re-lock a locked database every ten minutes. It's not ideal, but to be honest it should work fine.


> The following works to auto-lock the DB after 10 minutes of idleness. It should work on any wlroots-based compositor.
> 
> ```shell
> #!/bin/sh
> 
> while true; do
>   wayidle -t 600
>   gdbus call --session --dest org.keepassxc.KeePassXC.MainWindow --object-path /keepassxc --method org.keepassxc.KeePassXC.MainWindow.lockAllDatabases
>   # TODO: wait for signal indicating that the db is unlocked again.
> done
> ```

Why don't you just start the script when the screensaver was unlocked/the user logged in? Based on the presented use cases in this thread I'd think this system idle timeout will always be hit some time _after_ the screen locked, which would make a screen unlock the perfect time to restart it. In case your Keepass was never unlocked, it would be a no-op just as @droidmonkey stated. Still it wouldn't wake your system every x minutes to do basically nothing.

I still think having this tied to Keepass itself would be the cleaner and more universal approach.
Alternatively, if FIDO2 support gets merged in, some sort of quick unlock (ideally tied to the system/screensaver unlock) could mitigate this whole issue completely. Thinking of it, that should even be more secure than an independent timeout could ever be. Such a thing should probably be discussed in a separate issue though.

**Comment by WhyNotHugo** [comment]
> Why don't you just start the script when the screensaver was unlocked/the user logged in?

There is no correlation between screenlocker and KPXC being locked; KPXC will be auto-locked even if the screenlocker is inhibited. The KPXC timeout is not necessarily longer than the screenlocker timeout either (this particular value varies depending on host and other conditions).

> I still think having this tied to Keepass itself would be the cleaner and more universal approach.

Most Wayland compositors use a dedicated protocol, ext_idle_notifier_v1. I haven't found any indication of Qt actually exposing the interface in any way. I'm not sure how feasible it is to implement this without explicit Qt support.

Mutter specifically doesn't support ext_idle_notifier_v1 and uses its a bespoke D-Bus interface. This might be easier to implement, but will only work on a single compositor.

Generally, I think that providing the right hooks for external processes to handle this allows better per-desktop integration.

**Comment by jedinjapan** [comment]
This is a feature in the main Keepass 2.XX client that is invaluable for certain work scenarios, where you may remain at the workstation, actively engaged elsewhere in a application or browser, but need to pull up KeePassXC in order to fill in an application window or when credentials have been invalidated on a timer. It can also reduce repeated password re-entry from users actively using a system and reduce frustration with utilizing a password database.

Having this based on overall keyboard activity would mean that stepping away from the system for 5 minutes can lock the password vault (most critical - data can be read in plaintext by anyone) without locking the entire desktop (less critical - cookies and session token theft are a magnitude more difficult for average persons).

Additionally, if the user's database contains a mixture of personal/professional information, they may wish to lock the database more aggressively than they wish to lock the system itself. The option to trigger on screensaver is a workaround, but having this feature to being the application more in parity with KeePass 2.XX would be greatly appreciated.

**Comment by JohnLGalt** [comment]
I get around this by not letting the database auto lock itself, but having it lock when the system locks - and then locking my system when I get up.

Of course, I've made it simpler with Dynamic Lock using my phone, so my system autolocks when I take my phone far enough away, but still. It works fine in Windows.

Is there something similar that can be done in MacOS and Linux world?

**Comment by WhyNotHugo** [comment]
The main issue on Linux is that Qt doesn't expose ext_idle_notifier_v1 in any way (nor its X11 equivalent).

Using an external script to lock KeePassXC has worked for me, but isn't friendly for those who want a GUI approach.

Brainstorming: a separate platform-specific binary could be included, one which connects to the compositor (without Qt) and waits for the idle timeout. KPXC can spawn this helper itself, and make it controllable via the GUI.

**Comment by SynapseScribe** [comment]
WhyNot implement for windows only for now

**Comment by JohnLGalt** [comment]
> The main issue on Linux is that Qt doesn't expose ext_idle_notifier_v1 in any way (nor its X11 equivalent).
> 
> Using an external script to lock KeePassXC has worked for me, but isn't friendly for those who want a GUI approach.
> 
> Brainstorming: a separate platform-specific binary could be included, one which connects to the compositor (without Qt) and waits for the idle timeout. KPXC can spawn this helper itself, and make it controllable via the GUI.

So, let me brainstorm also. 

Don't *nix systems also allow you to 'lock' a system (ostensibly also requiring authentication to gain access again) after a time period or on demand? I haven't used *nix daily for a while, though I did just use the Gentoo LiveDVD for a clean install of Windows, and I did see there was a function built in to the KDE Plasma that allowed for locking the system (even allowing me to set it to use Win key + L), but I didn't explore it in much detail beyond seeing it and changing the hotkey. I had a system to re-image.

If such a 'lock' on a system would work, I think KPXC then has the functionality to lock the database when the system is being locked, (unless, of course, this is all tied to that one ext_idle_notified_v1, which, as a system function that can be manually enabled at any given moment, it probably should not be).

Just some food for thought - unless I'm wrong, and that particular `Lock database when session is locked or lid is closed` setting is only available on Windows....

**Comment by WhyNotHugo** [comment]
@JohnLGalt: What you refer to as "locking the system" is the  compositor (a.k.a.: display server) itself being locked. The `ext_idle_notified_v1` protocol extension is the mechanism used by clients to request that the compositor notifies them after N seconds of the session being idle.

The main issue is that Qt itself doesn't expose this to application developers, so it's non-trivial to use this from a Qt application. The Qt issue tracker has no mentions of `ext_idle_notified_v1`, so maybe it's worth asking if they're willing to expose this in some way.

**Comment by droidmonkey** [comment]
There should not be any need to listen for that notice as long as we can check for any indicators of system "usage" such as mouse movement, keyboard presses, etc.

**Comment by WhyNotHugo** [comment]
> There should not be any need to listen for that notice as long as we can check for any indicators of system "usage" such as mouse movement, keyboard presses, etc.

Clients aren't allowed to snoop on the keyboard events sent to other clients. You can't check for this kind of indicator without elevated privileges. Pretty sure that you'd need to be root to do this.

There's not even a privileged Wayland extension to do that. `ext_idle_notified_v1` exists for exactly this use case: to be notified after a certain period of idleness.
